p2p/protocol/holepunch/filter.go

```diff
@@ -0,0 +1,27 @@
+package holepunch
+
+import (
+	"github.com/libp2p/go-libp2p/core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// WithAddrFilter is a Service option that enables multiaddress filtering.
+// It allows to only send a subset of observed addresses to the remote
+// peer. E.g., only announce TCP or QUIC multi addresses instead of both.
+// It also allows to only consider a subset of received multi addresses
+// that remote peers announced to us.
+// Theoretically, this API also allows to add multi addresses in both cases.
+func WithAddrFilter(maf AddrFilter) Option {
```

Comment:
I'm inconsistently referring to the multiaddress filter as `AddrFilter` or just `filter` (see L16) 🤷‍♂️